File: app/src/main/java/com/gianessi/stargazers/adapters/UsersAdapter.java
package com.gianessi.stargazers.adapters;

import android.support.annotation.NonNull;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import com.bumptech.glide.Glide;
import com.bumptech.glide.request.RequestOptions;
import com.gianessi.stargazers.R;
import com.gianessi.stargazers.listeners.OnUserSelectedListener;
import com.gianessi.stargazers.models.User;

import java.util.List;

public class UsersAdapter extends RecyclerView.Adapter<RecyclerView.ViewHolder> {

    private static final int VIEW_TYPE_ITEM = 0;
    private static final int VIEW_TYPE_PROGRESS = 1;

    private final List<User> users;
    private OnUserSelectedListener listener;

    public UsersAdapter(List<User> users) {
        this.users = users;
    }

    public void setListener(OnUserSelectedListener listener) {
        this.listener = listener;
    }

    public void removeListener(){
        this.listener = null;
    }

    @Override
    public int getItemViewType(int position) {
        return this.users.get(position) == null ? VIEW_TYPE_PROGRESS : VIEW_TYPE_ITEM;
    }

    @NonNull
    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        if(viewType == VIEW_TYPE_ITEM) {
            View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.item_user, parent, false);
            return new UserViewHolder(view);
        }else{
            View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.item_progress, parent, false);
            return new ProgressViewHolder(view);
        }
    }

    @Override
    public void onBindViewHolder(@NonNull RecyclerView.ViewHolder viewHolder, int position) {
        if(viewHolder instanceof UserViewHolder){
            UserViewHolder userVH = (UserViewHolder) viewHolder;
            final User user = this.users.get(position);
            userVH.bind(user);
            userVH.itemView.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View view) {
                    if(listener != null)
                        listener.onUserSelected(user);
                }
            });
        }
    }

    @Override
    public int getItemCount() {
        return this.users.size();
    }

    private static class UserViewHolder extends RecyclerView.ViewHolder {

        private TextView usernameTxt;
        private ImageView avatarImg;

        private UserViewHolder(View view) {
            super(view);
            usernameTxt = view.findViewById(R.id.user_username_txt);
            avatarImg = view.findViewById(R.id.user_avatar_img);
        }

        private void bind(User user){
            Glide.with(this.itemView.getContext()).load(user.getAvatarUrl()).apply(RequestOptions.circleCropTransform()).into(avatarImg);
            this.usernameTxt.setText(user.getUsername());
        }
    }

    private static class ProgressViewHolder extends RecyclerView.ViewHolder {
        private ProgressViewHolder(View view) {
            super(view);
        }
    }

}

File: app/src/main/java/com/gianessi/stargazers/adapters/UsersAdapter.kt
package com.gianessi.stargazers.adapters

import androidx.recyclerview.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView

import com.bumptech.glide.Glide
import com.bumptech.glide.request.RequestOptions
import com.gianessi.stargazers.R
import com.gianessi.stargazers.listeners.OnUserSelectedListener
import com.gianessi.stargazers.models.User

private const val VIEW_TYPE_ITEM = 0
private const val VIEW_TYPE_PROGRESS = 1

class UsersAdapter(private val users: List<User?>) : androidx.recyclerview.widget.RecyclerView.Adapter<androidx.recyclerview.widget.RecyclerView.ViewHolder>() {

    var listener: OnUserSelectedListener? = null

    override fun getItemViewType(position: Int): Int {
        return if (this.users[position] == null) VIEW_TYPE_PROGRESS else VIEW_TYPE_ITEM
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): androidx.recyclerview.widget.RecyclerView.ViewHolder {
        return if (viewType == VIEW_TYPE_ITEM) {
            val view = LayoutInflater.from(parent.context).inflate(R.layout.item_user, parent, false)
            UserViewHolder(view)
        } else {
            val view = LayoutInflater.from(parent.context).inflate(R.layout.item_progress, parent, false)
            ProgressViewHolder(view)
        }
    }

    override fun onBindViewHolder(viewHolder: androidx.recyclerview.widget.RecyclerView.ViewHolder, position: Int) {
        (viewHolder as? UserViewHolder)?.let { userViewHolder ->
            val user = this.users[position] ?: return
            userViewHolder.bind(user)
            viewHolder.itemView.setOnClickListener { _ ->
                if (listener != null)
                    listener!!.onUserSelected(user)
            }
        }
    }

    override fun getItemCount(): Int {
        return this.users.size
    }

    inner class UserViewHolder (view: View) : androidx.recyclerview.widget.RecyclerView.ViewHolder(view) {

        private val usernameTxt: TextView = view.findViewById(R.id.user_username_txt)
        private val avatarImg: ImageView = view.findViewById(R.id.user_avatar_img)

        fun bind(user: User) {
            Glide.with(this.itemView.context).load(user.avatarUrl).apply(RequestOptions.circleCropTransform()).into(avatarImg)
            this.usernameTxt.text = user.username
        }
    }

    inner class ProgressViewHolder (view: View) : androidx.recyclerview.widget.RecyclerView.ViewHolder(view)

}
